File: tools/dsl-harness/kernels/triton/zig/moe_align_block_size.zig
//! Registration of `moe_align_block_size_kernel` for the harness.
//!
//! Re-exports the production Kernel from `zml.moe.triton_kernels`
//! (`MoeAlignBlockSize`). One default sweep mirroring the dump config in
//! `examples/triton_emitter/kernels_zig.zig`.
//!
//! In-place kernel: the 4 `tt.func` "output" pointers alias the 4 mutable
//! input buffers (`sorted_token_ids`, `expert_ids`, `num_tokens_post_pad`,
//! `cumsum`), so the XLA driver passes `output_operand_aliases` accordingly.

const std = @import("std");

const harness = @import("harness");
const zml = @import("zml");
const ops = zml.ops;
const Tensor = zml.Tensor;

pub const Kernel = zml.moe.triton_kernels.MoeAlignBlockSize.Kernel;

pub const SWEEPS: []const harness.Sweep(Kernel.Config) = &.{
    .{ .label = "default", .cfg = .{
        .numel = 1024,
        .num_experts = 8,
        .padded_num_experts = 8,
        .max_num_tokens_padded = 2048,
        .max_num_m_blocks = 32,
        .block_size_m = 64,
        .hist_block = 64,
    } },
};

// =============================================================================
// XLA-pipeline driver. 5 inputs + 4 outputs (out0..out3 aliasing inputs 1..4)
// → 9 `tt.func` params. Shapes match the default sweep; XLA only runs the
// codegen pipeline.
// =============================================================================

const NUMEL: i64 = 1024;
const NUM_EXPERTS: i64 = 8;
const MAX_NUM_TOKENS_PADDED: i64 = 2048;
const MAX_NUM_M_BLOCKS: i64 = 32;

threadlocal var active_ttir: [:0]const u8 = "";

pub fn setActiveTtir(ttir: [:0]const u8) void {
    active_ttir = ttir;
}

pub fn forward(
    topk_ids: Tensor,
    sorted_token_ids: Tensor,
    expert_ids: Tensor,
    num_tokens_post_pad: Tensor,
    cumsum: Tensor,
) struct { Tensor, Tensor, Tensor, Tensor } {
    const out = ops.triton(
        .{ topk_ids, sorted_token_ids, expert_ids, num_tokens_post_pad, cumsum },
        .{ sorted_token_ids.shape(), expert_ids.shape(), num_tokens_post_pad.shape(), cumsum.shape() },
        .{
            .name = Kernel.name,
            .ir = active_ttir,
            .grid = .{ 2, 1, 1 },
            .num_warps = 8,
            .num_stages = 1,
            .output_operand_aliases = &.{
                .{ .output_index = 0, .operand_index = 1 },
                .{ .output_index = 1, .operand_index = 2 },
                .{ .output_index = 2, .operand_index = 3 },
                .{ .output_index = 3, .operand_index = 4 },
            },
        },
    );
    return .{ out[0], out[1], out[2], out[3] };
}

pub fn args() std.meta.ArgsTuple(@TypeOf(forward)) {
    return .{
        Tensor.init(.{NUMEL}, .i32), // topk_ids_ptr
        Tensor.init(.{MAX_NUM_TOKENS_PADDED}, .i32), // sorted_token_ids_ptr
        Tensor.init(.{MAX_NUM_M_BLOCKS}, .i32), // expert_ids_ptr
        Tensor.init(.{1}, .i32), // num_tokens_post_pad_ptr
        Tensor.init(.{NUM_EXPERTS + 1}, .i32), // cumsum_ptr
    };
}

File: tools/dsl-harness/kernels/triton/py/moe_align_block_size.py
"""Python reference for `moe_align_block_size_kernel`."""

from __future__ import annotations

from typing import Any, Dict, Tuple

from triton_helpers import fake as _t
from moe import moe_align_block_size_kernel  # type: ignore

__all__ = ["moe_align_block_size_kernel", "build_args"]


def build_args(cfg: Dict[str, Any]) -> Tuple[list, Dict[str, Any]]:
    numel = int(cfg.get("numel", 1024))
    num_experts = int(cfg.get("num_experts", 8))
    padded_num_experts = int(cfg.get("padded_num_experts", num_experts))
    max_num_tokens_padded = int(cfg.get("max_num_tokens_padded", 2048))
    max_num_m_blocks = int(cfg.get("max_num_m_blocks", 32))
    block_size_m = int(cfg.get("block_size_m", 64))
    hist_block = int(cfg.get("hist_block", 64))

    args = [
        _t("i32", numel),                       # topk_ids_ptr
        _t("i32", max_num_tokens_padded),       # sorted_token_ids_ptr
        _t("i32", max_num_m_blocks),            # expert_ids_ptr
        _t("i32", 1),                           # num_tokens_post_pad_ptr
        _t("i32", num_experts + 1),             # cumsum_ptr
        block_size_m,                           # BLOCK_SIZE_M (constexpr)
        numel,                                  # NUMEL (constexpr)
        num_experts,                            # NUM_EXPERTS (constexpr)
        padded_num_experts,                     # PADDED_NUM_EXPERTS (constexpr)
        max_num_tokens_padded,                  # MAX_NUM_TOKENS_PADDED (constexpr)
        max_num_m_blocks,                       # MAX_NUM_M_BLOCKS (constexpr)
        hist_block,                             # HIST_BLOCK (constexpr)
        _t("i32", 1),                           # out0_ptr
        _t("i32", 1),                           # out1_ptr
        _t("i32", 1),                           # out2_ptr
        _t("i32", 1),                           # out3_ptr
    ]
    return args, {}
